Apply peerpod related MachineConfigs

bpradipt opened the pull request:
Peer pod enablement requires creating a new kata runtime class and providing the configuration.toml. In OpenShift we need to create MachineConfig objects to create these files.
The required MachineConfig yaml files are bundled within the controller-manager image and the manager process reads the yaml and creates the required MachineConfig objects. This gives us the flexibility to keep updating the yaml as needed without hardcoding the struct in the code

Fixes: #[Kata-2130](https://issues.redhat.com//browse/Kata-2130)

<!--
If this is a bug fix, make sure your description includes "Fixes: #xxxx", or
"Closes: #xxxx"

Please provide the following information:
-->
**- Description of the problem which is fixed/What is the use case**
PeerPod enablement requires creating configuration for a new runtimeclass and a new kata configuration.toml.
This PR enables creation of the required configs via the operator
**- What I did**
Added the required MachineConfig yamls to the controller-manager image and used the same in the code
to create the required configurations

**- How to verify it**
1. Create a KataConfig with `enablePeerPods: true`
2. Wait for KataConfig to be created successfully and `kata-oc` MCP to be in Updated state. 
3. Login the worker node and check for `/etc/crio/crio.conf.d/50-kata-remote-cc` and `/opt/kata/configuration-remote.toml`
4. These files are created by the MachineConfigs

**- Description for the changelog**
<!--
Write a short (one line) summary that describes the changes in this
pull request for inclusion in the changelog:
-->
Add peerpod related MachineConfigs

pmores commented on config/manager/kustomization.yaml:
```diff
@@ -12,5 +12,4 @@ apiVersion: kustomize.config.k8s.io/v1beta1
 - name: controller
-  newName: quay.io/openshift_sandboxed_containers/openshift-sandboxed-containers-operator
-  newTag: 1.3.1
+  newName: quay.io/bpradipt/openshift-sandboxed-containers-operator
```
This looks like a local modification, I guess it shouldn't be part of this PR?

bpradipt replied:
Right.

pmores commented on config/manager/manager.yaml:
```diff
@@ -62,7 +62,7 @@ spec:
           value: quay.io/openshift_sandboxed_containers/openshift-sandboxed-containers-monitor:latest
         - name: SANDBOXED_CONTAINERS_EXTENSION
-          value: kata-containers
+          value: sandboxed-containers
```
This too looks like a local modification, I guess it shouldn't be part of this PR?

bpradipt replied:
Right

pmores commented on config/peerpods/mc-50-crio-config.yaml:
```diff
@@ -2,7 +2,8 @@ apiVersion: machineconfiguration.openshift.io/v1
   labels:
-    machineconfiguration.openshift.io/role: worker
+    machineconfiguration.openshift.io/role: kata-oc
+    app: example-kataconfig
```
Does this belong upstream? (A honest question, admittedly I'm not sure how the `app` label is used.)

bpradipt replied:
It's a placeholder. This is not used by the MCO. The `app` label is also set for sandboxed extension - https://github.com/openshift/sandboxed-containers-operator/blob/main/controllers/openshift_controller.go#L483 
I think I should change it to the peerpodconfig name

pmores commented on controllers/openshift_controller.go:
```diff
@@ -936,6 +952,7 @@ func (r *KataConfigOpenShiftReconciler) processKataConfigInstallRequest() (ctrl.
 	}
 
 	// peer pod enablement
+	// This should be done after kata-oc creation and kata installation
```
Wouldn't this be executed _before_ kata-oc creation on the first reconciliation after a KataConfig instance is created?

bpradipt replied:
This is an incorrect comment. I'll remove it.

pmores commented on controllers/openshift_controller.go:
```diff
@@ -1727,14 +1742,119 @@ func (r *KataConfigOpenShiftReconciler) enablePeerPods() error {
+	peerpodconf := v1alpha1.PeerPodConfig{
+		TypeMeta: metav1.TypeMeta{},
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      "peerpodconfig-example",
```
Just a nitpick but we might want to use another name in published code than "-example"...

bpradipt replied:
Sure, I'll change it

pmores reviewed (approved):
Looks good to me, thanks Pradipta!

snir911 commented on controllers/openshift_controller.go:
```diff
@@ -1727,14 +1742,119 @@ func (r *KataConfigOpenShiftReconciler) enablePeerPods() error {
+func (r *KataConfigOpenShiftReconciler) createMcFromFile(mcFileName string) error {
+	yamlData, err := readMachineConfigYAML(mcFileName)
+	if err != nil {
+		r.Log.Error(err, "Unable to read MachineConfigYaml ", "mcFileName", mcFileName)
```
nit: generally speaking, why not returning formatted error instead

bpradipt replied:
The `err` is already formatted. Just logging it here for convenience.

snir911 reviewed (approved):
didn't try but LGTM overall, thanks!

jensfr reviewed (approved):
Lgtm, thank you pradipta!

gkurz commented on controllers/openshift_controller.go:
```diff
@@ -1727,14 +1742,119 @@ func (r *KataConfigOpenShiftReconciler) enablePeerPods() error {
 
-	err = r.Client.Create(context.TODO(), &peerpodconf)
+	err := r.Client.Create(context.TODO(), &peerpodconf)
 	if k8serrors.IsAlreadyExists(err) {
```
Does this mean that the two MCs and the runtime class are assumed to be created already if the peer pod config is already there ? I'm asking because if a previous call to this function created the peer pod config but not all the other objects it looks like it will never try to create them again.

bpradipt replied:
Right, it won't create the objects if peerpodconfig already exists.  Do you think it's too restrictive and better to handle this case?

gkurz replied:
I'd tend to say yes but since this code predates the PR I guess it can be addressed separately.

bpradipt replied:
@gkurz I went ahead and handled the case of existing peerpodconfig.

bpradipt commented:
In this version addressed the comments from @gkurz, especially switching to Log.Info instead of Log.Error and handling the case when peerpodconfig might exist

gkurz reviewed (approved):
LGTM. Thanks @bpradipt !
